### PR TITLE
Deleting clients creates MLS remove proposals

### PR DESCRIPTION
**Pull request by smatting**
## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/galley/test/integration/API/MLS.hs [hunk 1]
```diff
@@ -1630,3 +1632,81 @@ testBackendRemoveProposalLocalConvRemoteLeaver = do
         for_ bobClients $ \(_, ref) ->
           WS.assertMatch_ (5 # WS.Second) wsA $
             wsAssertBackendRemoveProposal bob qcnv ref
+
+-- for_ kprefs $ \kp ->
+--   WS.assertMatch_ (5 # WS.Second) wsA $ \notification ->
+--     void $
+--       wsAssertBackendRemoveProposal (pUserId bob) conversation kp notification
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs]: I suppose this has been removed in subsequent commits, but this is just a reminder in case it's not the case.

File: services/galley/test/integration/API/MLS.hs [hunk 2]
```diff
@@ -1630,3 +1632,81 @@ testBackendRemoveProposalLocalConvRemoteLeaver = do
...
+testBackendRemoveProposalLocalConvLocalClient :: TestM ()
+testBackendRemoveProposalLocalConvLocalClient = do
+  [alice, bob] <- createAndConnectUsers [Nothing, Nothing]
+
+  runMLSTest $ do
+    [alice1, bob1, bob2] <- traverse createMLSClient [alice, bob, bob]
+    traverse_ uploadNewKeyPackage [bob1, bob2]
+    (_, qcnv) <- setupMLSGroup alice1
+    void $ createAddCommit alice1 [bob] >>= sendAndConsumeCommit
+    [(bobA, kpBobA), (bobB, _kpBobB)] <- getClientsFromGroupState alice1 bob
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs:1615]: Just for readability purposes, I propose to add assertions `bobA == bob1` and `bobB == bob2`.

File: services/galley/test/integration/API/MLS.hs [hunk 3]
```diff
@@ -1630,3 +1632,81 @@ testBackendRemoveProposalLocalConvRemoteLeaver = do
...
+
+    let mock req = case frRPC req of
+          "on-conversation-updated" -> pure (Aeson.encode ())
+          "on-new-remote-conversation" -> pure (Aeson.encode EmptyResponse)
+          "get-mls-clients" ->
+            pure
+              . Aeson.encode
+              . Set.fromList
+              . map (flip ClientInfo True)
+              . toList
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs:1659]: `toList` on a list is an identity function, right?

File: services/galley/test/integration/API/MLS.hs [hunk 4]
```diff
@@ -752,6 +743,7 @@ testRemoteAppMessage = do
           "on-conversation-updated" -> pure (Aeson.encode ())
           "on-new-remote-conversation" -> pure (Aeson.encode EmptyResponse)
           "on-mls-message-sent" -> pure (Aeson.encode EmptyResponse)
+          "mls-welcome" -> pure (Aeson.encode EmptyResponse)
```
Comment (mdimjasevic) [inline services/galley/test/integration/API/MLS.hs:722]: Why is this needed in existing test cases? I'd assume the endpoint isn't called, otherwise the catch-all clause would trigger and we'd get an error.

That said, is is needed in the tests introduced by this PR?

**Review by mdimjasevic — COMMENTED** [review]
Overall this looks good, but I have some questions and comments that are inlined.

File: services/galley/src/Galley/API/Clients.hs [hunk 5]
```diff
@@ -61,9 +85,71 @@ addClientH (usr ::: clt) = do
...
+        goConvs range newCids lusr
+
+    removeRemoteMLSClients :: Range 1 1000 [Remote ConvId] -> Sem r ()
+    removeRemoteMLSClients cids = do
+      for_ (bucketRemote (fromRange cids)) $ \remoteConvs -> do
+        runFederatedEither remoteConvs (rpc (ClientRemovedRequest usr cid (tUnqualified remoteConvs)))
+          >>= logAndIgnoreError "Error in onConversationUpdated call" usr
+
+    logAndIgnoreError message usr' res = do
+      case res of
```
[the remark this answers is not in the document]
Comment (smatting) [reply services/galley/src/Galley/API/Clients.hs:143]: ah nice, didn't know about this one

File: services/galley/src/Galley/API/Clients.hs [hunk 6]
```diff
@@ -61,9 +85,71 @@ addClientH (usr ::: clt) = do
...
+        let nextState = mtpPagingState page
+            nextQuery = GetPaginatedConversationIds (Just nextState) range
+        newCids <- Query.conversationIdsPageFrom lusr nextQuery
+        goConvs range newCids lusr
+
+    removeRemoteMLSClients :: Range 1 1000 [Remote ConvId] -> Sem r ()
+    removeRemoteMLSClients cids = do
+      for_ (bucketRemote (fromRange cids)) $ \remoteConvs -> do
+        runFederatedEither remoteConvs (rpc (ClientRemovedRequest usr cid (tUnqualified remoteConvs)))
+          >>= logAndIgnoreError "Error in onConversationUpdated call" usr
```
[the remark this answers is not in the document]
Comment (smatting) [reply services/galley/src/Galley/API/Clients.hs:140]: Yes, it is. `rpc` is bound to `fedClient @'Galley @"on-client-removed"`. The reason we define this name outside is because creating the servant client `rpc` is costly apparently

File: services/galley/test/integration/API/MLS.hs (hunk 3 above)
Comment (smatting) [reply services/galley/test/integration/API/MLS.hs:1659]: Right, it was a set before I rebased this PR

File: services/galley/test/integration/API/MLS.hs (hunk 4 above)
Comment (smatting) [reply services/galley/test/integration/API/MLS.hs:722]: Errors get thrown actually when it's not implemented, however the tests still pass.

File: services/galley/src/Galley/API/Clients.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply services/galley/src/Galley/API/Clients.hs:140]: I am not sure I follow. I am asking if instead of:
```haskell
  logAndIgnoreError "Error in onConversationUpdated call" usr
```
it should have been:
```haskell
  logAndIgnoreError "Error in onClientRemoved call" usr
```
instead.

File: services/galley/test/integration/API/MLS.hs (hunk 2 above)
Comment (smatting) [reply services/galley/test/integration/API/MLS.hs:1615]: This assumption can't be made. I went with a solution involving `find`

**Comment by smatting** [comment]
@mdimjasevic Thanks for the review! I followed all your suggetions in https://github.com/wireapp/wire-server/pull/2674/commits/85575f030ce666b614aebe6da49846b59077f281

File: services/galley/test/integration/API/MLS.hs (hunk 2 above)
Comment (mdimjasevic) [reply services/galley/test/integration/API/MLS.hs]: I am not sure I follow. I made this change locally on top of the PR:
```diff
diff --git a/services/galley/test/integration/API/MLS.hs b/services/galley/test/integration/API/MLS.hs
index 0927979dc..d562d1f2c 100644
--- a/services/galley/test/integration/API/MLS.hs
+++ b/services/galley/test/integration/API/MLS.hs
@@ -1613,6 +1613,8 @@ testBackendRemoveProposalLocalConvLocalClient = do
     (_, qcnv) <- setupMLSGroup alice1
     void $ createAddCommit alice1 [bob] >>= sendAndConsumeCommit
     [(bobA, kpBobA), (bobB, _kpBobB)] <- getClientsFromGroupState alice1 bob
+    liftIO $ assertEqual "bob1 /= bobA" bob1 bobA
+    liftIO $ assertEqual "bob2 /= bobB" bob2 bobB
```
and the test passes. These are assumptions I read the test with, and these are just codified versions of the assumptions.

**Review by pcapriotti — APPROVED** [review]
LGTM

File: services/galley/test/integration/API/MLS.hs (hunk 2 above)
Comment (smatting) [reply services/galley/test/integration/API/MLS.hs]: What I mean is that the order of `getClientsFromGroupState` is not guaranteed to be the order of generation. It might work, but I wouldn't trust it. For the test it's arbitraty whether `bob1` or `bob2` is deleted. But in any case with the latest changes I can pick `bob1` and have a less confusing test

Comment (mdimjasevic) [reply services/galley/test/integration/API/MLS.hs]: > What I mean is that the order of `getClientsFromGroupState` is not guaranteed to be the order of generation. It might work, but I wouldn't trust it. For the test it's arbitraty whether `bob1` or `bob2` is deleted. But in any case with the latest changes I can pick `bob1` and have a less confusing test

Oh, thanks for clarifying this! I wish we had `Map ClientIdentity KeyPackageRef` in the return type of `getClientsFromGroupState` instead of `[(ClientIdentity, KeyPackageRef)]`, which would clearly tell there's no order to expect among the pairs of associated values.
